### PR TITLE
OCPNODE-1330 : Set the CGroups version explicitly to "v1"

### PR DESCRIPTION
**Pull request by sairameshv**
1. The CGroups version on an OCP cluster can be altered by altering the `nodes.config` resource's `spec.cgroupMode` field
2. It is very likely that the CGroups version would be defaulting to "v2" starting from OCP-4.13 on the underlying RCOS nodes.
3. To avoid unexpected complications, this code explicitly sets the CGroups version to "v1" on the newer OCP-4.13 versions
4. Also improved the bootstrap test cases with a few special case scenarios.

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

**- What I did**
Added code to explicitly set the `cgroupMode` to "v1", if found empty, in the newer OCP clusters(4.13)
**- How to verify it**
1. Deploy an OCP cluster (version = 4.13)
2. Debug into any of the nodes and observe the `CGroupsV1` configuration by executing the following command
```
sh-4.4# stat -c %T -f /sys/fs/cgroup/
tmpfs
```
Also observe the kernelArguments in the newly rendered machine config objects for the master and worker pools as follows:
```
kernelArguments:
  - systemd.unified_cgroup_hierarchy=0
  - systemd.legacy_systemd_cgroup_controller=1
```
**- Description for the changelog**

Explicitly set the `CGroups` configuration to `v1`, if found empty, in the newer version of OCP (4.13)

### CODE REVIEW

File: pkg/controller/kubelet-config/kubelet_config_nodes.go [hunk 1]
```diff
@@ -77,6 +77,11 @@ func (ctrl *Controller) syncNodeConfigHandler(key string) error {
 		err := fmt.Errorf("could not fetch Node: %w", err)
 		return err
 	}
+	// Set the Cgroups version explicitly to "v1"
+	if nodeConfig.Spec.CgroupMode == emptyInput {
```
Comment (yuqi-zhang) [inline pkg/controller/kubelet-config/kubelet_config_nodes.go:81]: can this be nil somehow? I forget how `omitempty` works, or does this check cover that already

File: pkg/controller/kubelet-config/kubelet_config_nodes.go [hunk 2]
```diff
@@ -77,6 +77,11 @@ func (ctrl *Controller) syncNodeConfigHandler(key string) error {
...
+		nodeConfig.Spec.CgroupMode = osev1.CgroupModeV1
+		ctrl.configClient.ConfigV1().Nodes().Update(context.TODO(), nodeConfig, metav1.UpdateOptions{})
```
Comment (yuqi-zhang) [inline pkg/controller/kubelet-config/kubelet_config_nodes.go:83]: Hmm, is it needed to update the config back? I see below you are updating the MachineConfig entry to have the right default, would that be enough?

This also could fail due to e.g. conflict writes, transient networking issues, etc. so if we dont need this maybe we just keep the node config intact (and empty) and just change MC values. WDYT?

File: pkg/controller/kubelet-config/kubelet_config_nodes.go [hunk 3]
```diff
@@ -270,8 +287,15 @@ func (ctrl *Controller) deleteNodeConfig(obj interface{}) {
 }
 
 func RunNodeConfigBootstrap(templateDir string, features *osev1.FeatureGate, cconfig *mcfgv1.ControllerConfig, nodeConfig *osev1.Node, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	if nodeConfig == nil {
```
Comment (yuqi-zhang) [inline pkg/controller/kubelet-config/kubelet_config_nodes.go:290]: For us to get into this function, there was an explicit `nodeConfig != nil` check, so maybe we don't need this?

**Review by yuqi-zhang — COMMENTED** [review]
Generally makes sense, some questions/comments below

File: pkg/controller/kubelet-config/kubelet_config_nodes.go (hunk 1 above)
Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:81]: `nodeConfig` can't be nil here as we return error [above](https://github.com/openshift/machine-config-operator/blob/c916a0fd9f410397741e9248c2e1038371872a06/pkg/controller/kubelet-config/kubelet_config_nodes.go#L78) if it is `nil`

File: pkg/controller/kubelet-config/kubelet_config_nodes.go (hunk 2 above)
Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:83]: Thought: As of now, we overwrite the CGroup version "V1" even if it is found to be "". This update makes sense here, but in future, if we don't need to overwrite, this update doesn't make sense (updating the same already existing Config Node object).
Hence, I would remove this step.

File: pkg/controller/kubelet-config/kubelet_config_nodes.go (hunk 3 above)
Comment (sairameshv) [reply pkg/controller/kubelet-config/kubelet_config_nodes.go:290]: This function would also be called from the test framework to test the bootstrap behavior when a `nil` `nodeConfig` is sent. Hence, this check has been added. However in the real cluster scenario, we are defaulting the node object if found `nil`

**Comment by yuqi-zhang** [comment]
Updates lgtm, do you know what the prow status

> Merging is blocked by issue 637.

Means here? I don't see a Jira issue 637 referenced anywhere

**Comment by yuqi-zhang** [comment]
Ah I see, this should be open against master branch, not `release-4.13`. Prow will auto fast forward the branch but otherwise you can't merge directly into `release-4.13`

**Comment by sairameshv** [comment]
> Ah I see, this should be open against master branch, not `release-4.13`. Prow will auto fast forward the branch but otherwise you can't merge directly into `release-4.13`

Oh, Okay. Thanks !!
I would raise it against the master branch and let you know.

**Comment by harche** [comment]
@sairameshv @yuqi-zhang the changes need to go only in 4.13. These changes should not be part of 4.14 in the future. 

cc @mrunalp
